conbench/entities/history.py

```diff
@@ -85,31 +89,170 @@ def get_history(case_id, context_id, hardware_hash, repo) -> List[tuple]:
+        history,
+        func.stddev(history.c.residual)
+        .over(
+            order_by=s.extract("epoch", history.c.timestamp),
+            range_=(None, -1),
+        )
+        .label("rolling_sd_excluding_this_commit"),
+    )
+
+    return history.all()
+
+
+def _calc_z_score(
+    data_point: float, unit: str, dist_mean: Optional[float], dist_sd: Optional[float]
+) -> Optional[float]:
+    """Calculate the z-score of a data point compared to a distribution."""
+    if dist_mean is not None and dist_sd is not None and dist_sd != 0:
+        z_score = (data_point - dist_mean) / dist_sd
+    else:
+        z_score = None
+
+    if z_score and _less_is_better(unit):
+        z_score = z_score * -1
+
+    return z_score
+
+
+def append_z_scores(benchmark_results: List[BenchmarkResult]):
+    """Populate the "z_score" attribute of each given BenchmarkResult, comparing it to
+    the distribution of BenchmarkResults that share the same hardware/case/context/repo,
+    ending with the most recent BenchmarkResult in the given BenchmarkResult's commit
+    ancestry on the default branch.
+
+    We choose a commit from the default branch as the end commit, because we probably
+    don't want "bad" commits early on in a PR to set artificially low z-scores later in
+    the PR.
+
+    Should not raise anything. If we can't find a z-score for some reason, the z_score
+    attribute will be None on that BenchmarkResult.
+    """
+    # {(case, context, hardware, repo, commit): (mean, sd)}
+    cached_stats: Dict[
+        Tuple[str, str, str, str, str],
+        Tuple[Optional[float], Optional[float]],
+    ] = {}
+
+    for benchmark_result in benchmark_results:
+        if benchmark_result.error or benchmark_result.mean is None:
+            benchmark_result.z_score = None
+            continue
+
+        baseline_run = benchmark_result.run.get_baseline_run(
+            on_default_branch=True,
+            case_id=benchmark_result.case_id,
+            context_id=benchmark_result.context_id,
+        )
+        if not baseline_run:
+            log.debug(
+                f"BenchmarkResult {benchmark_result.id} has no baseline run; "
+                "setting z_score = None"
+            )
+            benchmark_result.z_score = None
+            continue
+
+        case = benchmark_result.case_id
+        context = benchmark_result.context_id
+        hardware = benchmark_result.run.hardware.hash
+        repo = benchmark_result.run.commit.repository
+        baseline_sha = baseline_run.commit.sha
+
+        if (case, context, hardware, repo, baseline_sha) not in cached_stats:
+            # Fill in the cache with all the stats for this history
+            history = get_history(case, context, hardware, repo)
+            for res in history:
+                cached_stats[(case, context, hardware, repo, res.sha)] = (
+                    res.rolling_mean_including_this_commit,
+                    res.rolling_sd_including_this_commit,
```

Comment:
Does `rolling_sd_including_this_commit` include the current contender commit value in this calculation? I think yes, but want to double check.

Reply:
Yeah, it does. I think the way it's written now is confusing. I'm gonna rewrite it a little.

Reply:
Ok, I don't think that we want that to be how we calculate it, we probably want to use `rolling_sd_excluding_this_commit` since we don't want the contender to influence the distribution. With enough observations in history the impact here should be small, but we don't want something that is a big outlier to make it less likely we would detect that because we're including it in the distribution we're testing against.

Reply:
Oh I see what you were asking. Yeah sorry, this is what's confusing about how I wrote this code. We aren't including the contender commit in the distribution when calculating z-scores. We ARE including the baseline commit though, and this code is accessing the distribution AT the baseline commit, so we use `_including_this_commit` not `_excluding_this_commit`.

I'll rewrite this to make it more clear. :)

Reply:
great exchange, thanks